File: LCA_BST_Iterative.java
/*
## Problem 2: Lowest Common Ancestor of a Binary Search Tree 
(https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/)

Time Complexity :   O (h) 
Space Complexity :  O (h) 
Did this code successfully run on Leetcode :    Yes (235. Lowest Common Ancestor of a Binary Search Tree)
Any problem you faced while coding this :       No
 */
// Input: root = [6,2,8,0,4,7,9,null,null,3,5], p = 2, q = 8
// Output: 6
    
class LowestCommonAncestor {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        while(true){
            if(root.val < p.val && root.val < q.val)
                root = root.right;
            else if(root.val > p.val && root.val > q.val)
                root = root.left;
            else
                return root;
        }
    }
}

File: LCA_BST_Iterative.py
# Time complexity : O(n)
# Space complexity : O(1)
# Leetcode : Solved and submitted

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        # check for root if null
        if not root:
            return None
        
        # we continue until we reach the else part
        while True:
          # if both values are lesser, move to left sub-tree
            if root.val > p.val and root.val > q.val:
                root = root.left
          # if both values are greater, move to right sub-tree
            elif root.val < p.val and root.val < q.val:
                root = root.right
            else:
              # else root is the LCA
                return root
